lib/web/databases.go

```diff
@@ -387,13 +406,339 @@ func (h *Handler) sqlServerConfigureADScriptHandle(w http.ResponseWriter, r *htt
+			Roles:       s.req.DatabaseRoles,
+		},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.log.DebugContext(s.ctx, "Starting database interactive session")
+	if err := repl.Run(s.ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.log.DebugContext(s.ctx, "Database interactive session exited with success")
+	return nil
+}
+
+func (s *databaseInteractiveSession) Close() error {
+	s.replConn.Close()
+	return s.ws.Close()
+}
+
+// issueCerts performs the MFA (if required) and generate the user session
+// certificates.
+func (s *databaseInteractiveSession) issueCerts() (*tls.Certificate, *clientproto.RouteToDatabase, error) {
+	pk, err := keys.ParsePrivateKey(s.sctx.cfg.Session.GetTLSPriv())
+	if err != nil {
+		return nil, nil, trace.Wrap(err, "failed getting user private key from the session")
+	}
+
+	publicKeyPEM, err := keys.MarshalPublicKey(pk.Public())
+	if err != nil {
+		return nil, nil, trace.Wrap(err, "failed to marshal public key")
+	}
+
+	routeToDatabase := clientproto.RouteToDatabase{
+		Protocol:    s.db.GetProtocol(),
+		ServiceName: s.req.ServiceName,
+		Username:    s.req.DatabaseUser,
+		Database:    s.req.DatabaseName,
+		Roles:       s.req.DatabaseRoles,
+	}
+
+	certsReq := clientproto.UserCertsRequest{
+		TLSPublicKey:    publicKeyPEM,
+		Username:        s.sctx.GetUser(),
+		Expires:         s.sctx.cfg.Session.GetExpiryTime(),
+		Format:          constants.CertificateFormatStandard,
+		RouteToCluster:  s.site.GetName(),
+		Usage:           clientproto.UserCertsRequest_Database,
+		RouteToDatabase: routeToDatabase,
+	}
+
+	_, certs, err := client.PerformSessionMFACeremony(s.ctx, client.PerformSessionMFACeremonyParams{
+		CurrentAuthClient: s.clt,
+		RootAuthClient:    s.sctx.cfg.RootClient,
+		MFACeremony:       newMFACeremony(s.stream.WSStream, s.sctx.cfg.RootClient.CreateAuthenticateChallenge),
+		MFAAgainstRoot:    s.sctx.cfg.RootClusterName == s.site.GetName(),
+		MFARequiredReq: &clientproto.IsMFARequiredRequest{
+			Target: &clientproto.IsMFARequiredRequest_Database{Database: &routeToDatabase},
+		},
+		CertsReq: &certsReq,
+	})
+	if err != nil && !errors.Is(err, services.ErrSessionMFANotRequired) {
+		return nil, nil, trace.Wrap(err, "failed performing mfa ceremony")
+	}
+
+	if certs == nil {
+		certs, err = s.sctx.cfg.RootClient.GenerateUserCerts(s.ctx, certsReq)
+		if err != nil {
+			return nil, nil, trace.Wrap(err, "failed issuing user certs")
+		}
+	}
+
+	tlsCert, err := pk.TLSCertificate(certs.TLS)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return &tlsCert, &routeToDatabase, nil
+}
+
+func (s *databaseInteractiveSession) sendSessionMetadata() error {
+	sessionMetadataResponse, err := json.Marshal(siteSessionGenerateResponse{Session: session.Session{
+		// TODO(gabrielcorado): Have a consistent Session ID. Right now, the
+		// initial session ID returned won't be correct as the session is only
+		// initialized by the database server after the REPL starts.
```

Comment:
Can we skip sending the session id since we know it won't be correct anyway?

Reply:
Done. Once we have solved this this function can adequately return the session ID.